### ISSUE TITLE
GPU freezes when dcgm-exporter is SIGKILL'd

### ISSUE THREAD

**Issue by mac-chaffee** [NVIDIA/dcgm-exporter#84]
* GPU Type: A100
* Driver Version: 515.48.07
* OS: RHEL8 running Kubernetes
* dcgm-exporter version: 2.3.4-2.6.4-ubuntu20.04
* MIG: yes

If the dcgm-exporter is forcibly killed (either `kill -9`, takes to long to respond to SIGTERM and k8s SIGKILL's it, or an oomkill), it appears to cause my GPU to freeze. `nvidia-smi` hangs and no other processes are able to use the GPU until the server is restarted.

Since I also observe dcgm-exporter having a memory leak as noted in #340, that means SIGKILLs can be a regular occurrence. 

When the GPU is frozen, I see this in `dmesg`:

```
[Mon Jul 25 16:46:04 2022] NVRM: GPU Board Serial Number: 1565020013641
[Mon Jul 25 16:46:04 2022] NVRM: Xid (PCI:0000:3b:00): 120, pid='<unknown>', name=<unknown>, GSP Error: Task 1 raised error code 0x5 for reason 0x0 at 0x63f01ac (0 more errors skipped)
[Mon Jul 25 16:46:20 2022] NVRM: Xid (PCI:0000:3b:00): 119, pid='<unknown>', name=<unknown>, Timeout waiting for RPC from GSP! Expected function FREE (0x0 0x0).
[Mon Jul 25 16:46:20 2022] NVRM: Xid (PCI:0000:3b:00): 119, pid='<unknown>', name=<unknown>, Timeout waiting for RPC from GSP! Expected function FREE (0x0 0x0).
[Mon Jul 25 16:46:20 2022] NVRM: Xid (PCI:0000:3b:00): 119, pid='<unknown>', name=<unknown>, Timeout waiting for RPC from GSP! Expected function FREE (0x0 0x0).
[Mon Jul 25 16:46:20 2022] NVRM: Xid (PCI:0000:3b:00): 119, pid='<unknown>', name=<unknown>, Timeout waiting for RPC from GSP! Expected function FREE (0x0 0x0).
[Mon Jul 25 16:46:20 2022] NVRM: Xid (PCI:0000:3b:00): 119, pid='<unknown>', name=<unknown>, Timeout waiting for RPC from GSP! Expected function FREE (0x0 0x0).
[Mon Jul 25 16:46:20 2022] NVRM: GPU at PCI:0000:d8:00: GPU-d7099080-bc3c-6429-51d1-5825fdccd129
[Mon Jul 25 16:46:20 2022] NVRM: GPU Board Serial Number: 1565020014726
[Mon Jul 25 16:46:20 2022] NVRM: Xid (PCI:0000:d8:00): 119, pid='<unknown>', name=<unknown>, Timeout waiting for RPC from GSP! Expected function FREE (0x0 0x0).
[Mon Jul 25 16:46:20 2022] NVRM: Xid (PCI:0000:d8:00): 119, pid='<unknown>', name=<unknown>, Timeout waiting for RPC from GSP! Expected function FREE (0x0 0x0).
[Mon Jul 25 16:46:20 2022] NVRM: Xid (PCI:0000:d8:00): 119, pid='<unknown>', name=<unknown>, Timeout waiting for RPC from GSP! Expected function FREE (0x0 0x0).
[Mon Jul 25 16:46:20 2022] NVRM: Xid (PCI:0000:d8:00): 119, pid='<unknown>', name=<unknown>, Timeout waiting for RPC from GSP! Expected function FREE (0x0 0x0).
[Mon Jul 25 16:46:20 2022] NVRM: Xid (PCI:0000:d8:00): 119, pid='<unknown>', name=<unknown>, Timeout waiting for RPC from GSP! Expected function FREE (0x0 0x0).
[Mon Jul 25 16:46:20 2022] NVRM: Xid (PCI:0000:d8:00): 119, pid='<unknown>', name=<unknown>, Timeout waiting for RPC from GSP! Expected function FREE (0x0 0x0).
[Mon Jul 25 16:46:20 2022] NVRM: Xid (PCI:0000:d8:00): 120, pid='<unknown>', name=<unknown>, GSP Error: Task 1 raised error code 0x5 for reason 0x0 at 0x63f01ac (0 more errors skipped)
[Mon Jul 25 16:46:26 2022] NVRM: Xid (PCI:0000:3b:00): 119, pid='<unknown>', name=<unknown>, Timeout waiting for RPC from GSP! Expected function GSP_RM_CONTROL (0x20801348 0x410).
[Mon Jul 25 16:46:30 2022] NVRM: Xid (PCI:0000:3b:00): 119, pid='<unknown>', name=<unknown>, Timeout waiting for RPC from GSP! Expected function GSP_RM_ALLOC (0x0 0x6c).
[Mon Jul 25 16:46:30 2022] NVRM: Xid (PCI:0000:3b:00): 119, pid='<unknown>', name=<unknown>, Timeout waiting for RPC from GSP! Expected function GSP_RM_ALLOC (0x80 0x38).
[Mon Jul 25 16:46:34 2022] NVRM: Xid (PCI:0000:3b:00): 119, pid='<unknown>', name=<unknown>, Timeout waiting for RPC from GSP! Expected function GSP_RM_ALLOC (0x2080 0x4).
[Mon Jul 25 16:46:38 2022] NVRM: Xid (PCI:0000:3b:00): 119, pid='<unknown>', name=<unknown>, Timeout waiting for RPC from GSP! Expected function FREE (0x0 0x0).
[Mon Jul 25 16:46:42 2022] NVRM: Xid (PCI:0000:3b:00): 119, pid='<unknown>', name=<unknown>, Timeout waiting for RPC from GSP! Expected function FREE (0x0 0x0).
[Mon Jul 25 16:46:55 2022] NVRM: Xid (PCI:0000:3b:00): 119, pid='<unknown>', name=<unknown>, Timeout waiting for RPC from GSP! Expected function GSP_RM_CONTROL (0x20801348 0x410).
[Mon Jul 25 16:47:03 2022] NVRM: Xid (PCI:0000:3b:00): 119, pid='<unknown>', name=<unknown>, Timeout waiting for RPC from GSP! Expected function GSP_RM_ALLOC (0x0 0x6c).
[Mon Jul 25 16:47:03 2022] NVRM: Xid (PCI:0000:3b:00): 119, pid='<unknown>', name=<unknown>, Timeout waiting for RPC from GSP! Expected function GSP_RM_ALLOC (0x80 0x38).
[Mon Jul 25 16:47:11 2022] NVRM: Xid (PCI:0000:3b:00): 119, pid='<unknown>', name=<unknown>, Timeout waiting for RPC from GSP! Expected function GSP_RM_ALLOC (0x2080 0x4).
```

That Xid code appears to be undocumented: https://docs.nvidia.com/deploy/xid-errors/index.html

Also reported here: https://forums.developer.nvidia.com/t/a100-gpu-freezes-after-process-gets-oomkilled/221630

Any ideas what could be causing this?

**Comment by nikkon-dev** [comment]
@mac-chaffee,

Could you specify which metrics are enabled in the dcgm-exporter?

WBR,
Nik

**Comment by nikkon-dev** [comment]
Just for clarity - XID 119 stands for "GSP RPC Timeout".

**Comment by nikkon-dev** [comment]
@mac-chaffee,

I confirmed this is a known issue when DCP metrics (1001-1014) are enabled and the nv-hostengine process is killed. Unfortunately, we cannot just fix this on the DCGM side as a driver fix is needed.

**Comment by mac-chaffee** [comment]
Thanks for confirming!

Are you aware of any workarounds? Could I disable collection of DCP metrics somehow? Looks like I could also disable the GSP entirely, not sure of the ramifications of that: https://download.nvidia.com/XFree86/Linux-x86_64/510.39.01/README/gsp.html

**Comment by nikkon-dev** [comment]
@mac-chaffee,

If you do not need DCP metrics (DCGM_FI_PROF*), then disabling them would be the easiest way to mitigate this. You could specify another CSV file - default-counters.csv instead of dcp-metrics-included.csv or provide your CSV file or use ConfigMap instead.

To disable GSP Firmware, you need to specify `NVreg_EnableGpuFirmware=0` to the Nvidia driver via `/etc/modprobe.d/nvidia.conf` or manually.

**Comment by PlayMTL** [comment]
I have the same problem withUbuntu 20.04 and Nvidia driver 470 when using the DCGM Exporter, unfortunately the workarounds do not help me. Disabling GSP firmware doesn't work and disabling the metrics didn't work either. Are there any other solutions?

**Comment by vedantroy** [comment]
@PlayMTL did you ever manage to find a solution?

**Comment by PlayMTL** [comment]
Cause the Workaround wasn't working for me, i tried something new:

I switched the driver version from NVIDIA to 515.65 and disabled MIG. Since 24h I don't had anymore crashes / freezes.
My DCGM-Exporter Version is 2.4.6-2.6.10-ubuntu20.04

**Comment by zhangineer** [comment]
I have a similar issue, but I don't use DCGM at all.
Is this only related to GSP then?

```
Nov 16 01:40:56 hpe-host4 kernel: [ 1354.379821] OOM killer enabled.
Nov 16 01:40:56 hpe-host4 kernel: [ 1354.379822] Restarting tasks ... done.
Nov 16 01:40:56 hpe-host4 kernel: [ 1354.383996] PM: suspend exit
Nov 16 01:40:56 hpe-host4 kernel: [ 1354.384247] PM: suspend entry (s2idle)
Nov 16 01:40:56 hpe-host4 kernel: [ 1354.393806] Filesystems sync: 0.009 seconds
Nov 16 01:41:08 hpe-host4 kernel: [ 1354.433110] Freezing user space processes ... (elapsed 0.067 seconds) done.
Nov 16 01:41:08 hpe-host4 kernel: [ 1354.500190] OOM killer disabled.
Nov 16 01:41:08 hpe-host4 kernel: [ 1354.500192] Freezing remaining freezable tasks ... (elapsed 0.001 seconds) done.
Nov 16 01:41:08 hpe-host4 kernel: [ 1354.501870] printk: Suspending console(s) (use no_console_suspend to debug)
Nov 16 01:41:08 hpe-host4 kernel: [ 1360.506474] NVRM: GPU at PCI:0000:3c:00: GPU-94d4cdf3-c00f-14e9-ba5b-195d1be18c1f
Nov 16 01:41:08 hpe-host4 kernel: [ 1360.506484] NVRM: Xid (PCI:0000:3c:00): 119, pid='<unknown>', name=<unknown>, Timeout waiting for RPC from GSP! Expected function GSP_RM_CONTROL (0x20803031 0x4).
Nov 16 01:41:08 hpe-host4 kernel: [ 1366.510254] NVRM: Xid (PCI:0000:3c:00): 119, pid='<unknown>', name=<unknown>, Timeout waiting for RPC from GSP! Expected function UNLOADING_GUEST_DRIVER (0x0 0x0).
Nov 16 01:41:08 hpe-host4 kernel: [ 1366.562997] PM: pci_pm_suspend(): nv_pmops_suspend+0x0/0x30 [nvidia] returns -5
Nov 16 01:41:08 hpe-host4 kernel: [ 1366.563003] PM: dpm_run_callback(): pci_pm_suspend+0x0/0x150 returns -5
Nov 16 01:41:08 hpe-host4 kernel: [ 1366.563010] PM: Device 0000:3c:00.0 failed to suspend async: error -5
Nov 16 01:41:08 hpe-host4 kernel: [ 1366.563294] PM: Some devices failed to suspend, or early wake event detected
```

**Comment by zhangineer** [comment]
> I have a similar issue, but I don't use DCGM at all. Is this only related to GSP then? ( I am trying the workaround to disable the firmware, will report back result )
> 
> ```
> Nov 16 01:40:56 hpe-host4 kernel: [ 1354.379821] OOM killer enabled.
> Nov 16 01:40:56 hpe-host4 kernel: [ 1354.379822] Restarting tasks ... done.
> Nov 16 01:40:56 hpe-host4 kernel: [ 1354.383996] PM: suspend exit
> Nov 16 01:40:56 hpe-host4 kernel: [ 1354.384247] PM: suspend entry (s2idle)
> Nov 16 01:40:56 hpe-host4 kernel: [ 1354.393806] Filesystems sync: 0.009 seconds
> Nov 16 01:41:08 hpe-host4 kernel: [ 1354.433110] Freezing user space processes ... (elapsed 0.067 seconds) done.
> Nov 16 01:41:08 hpe-host4 kernel: [ 1354.500190] OOM killer disabled.
> Nov 16 01:41:08 hpe-host4 kernel: [ 1354.500192] Freezing remaining freezable tasks ... (elapsed 0.001 seconds) done.
> Nov 16 01:41:08 hpe-host4 kernel: [ 1354.501870] printk: Suspending console(s) (use no_console_suspend to debug)
> Nov 16 01:41:08 hpe-host4 kernel: [ 1360.506474] NVRM: GPU at PCI:0000:3c:00: GPU-94d4cdf3-c00f-14e9-ba5b-195d1be18c1f
> Nov 16 01:41:08 hpe-host4 kernel: [ 1360.506484] NVRM: Xid (PCI:0000:3c:00): 119, pid='<unknown>', name=<unknown>, Timeout waiting for RPC from GSP! Expected function GSP_RM_CONTROL (0x20803031 0x4).
> Nov 16 01:41:08 hpe-host4 kernel: [ 1366.510254] NVRM: Xid (PCI:0000:3c:00): 119, pid='<unknown>', name=<unknown>, Timeout waiting for RPC from GSP! Expected function UNLOADING_GUEST_DRIVER (0x0 0x0).
> Nov 16 01:41:08 hpe-host4 kernel: [ 1366.562997] PM: pci_pm_suspend(): nv_pmops_suspend+0x0/0x30 [nvidia] returns -5
> Nov 16 01:41:08 hpe-host4 kernel: [ 1366.563003] PM: dpm_run_callback(): pci_pm_suspend+0x0/0x150 returns -5
> Nov 16 01:41:08 hpe-host4 kernel: [ 1366.563010] PM: Device 0000:3c:00.0 failed to suspend async: error -5
> Nov 16 01:41:08 hpe-host4 kernel: [ 1366.563294] PM: Some devices failed to suspend, or early wake event detected
> ```

using the latest CUDA and drivers, i have not encountered any issues so far ( 24 hours + now )

**Comment by jaeeyoungkim** [comment]
this is driver issue . fixed in 535.216.01

https://docs.nvidia.com/datacenter/tesla/tesla-release-notes-535-216-01/index.html

<img width="1205" height="53" alt="Image" src="https://github.com/user-attachments/assets/00a3ab78-6618-4f28-9e2e-b09bf847f16f" />


This issue occurred in the A100 MIG environment, and it was resolved after upgrading the NVIDIA driver to version 535.216 or later.

There were a sufficient number of GPU workloads running — more than 50 pods.

The A100 node had 8 GPUs, and they were partitioned into 56 MIG instances.
When dcgm-exporter experienced an OOM kill or when dcgm-exporter was deleted, an XID 120 error occurred.
The XID error was found in the system logs on the affected node.

I have updated our environment to 535.274.02.
